quantum/rgb_matrix.c

```diff
@@ -0,0 +1,873 @@
+#include "TWIlib.h"
+#include <util/delay.h>
+#include <avr/interrupt.h>
+#include "progmem.h"
+#include "config.h"
+#include "eeprom.h"
+#include "lufa.h"
+#include <math.h>
+
+rgb_config_t rgb_matrix_config;
+
+#ifndef RGB_DISABLE_AFTER_TIMEOUT
+    #define RGB_DISABLE_AFTER_TIMEOUT 0
+#endif
+
+#ifndef RGB_DISABLE_WHEN_USB_SUSPENDED
+    #define RGB_DISABLE_WHEN_USB_SUSPENDED false
+#endif
+
+#ifndef EECONFIG_RGB_MATRIX
+    #define EECONFIG_RGB_MATRIX EECONFIG_RGBLIGHT
+#endif
+
+bool g_suspend_state = false;
+
+// Global tick at 20 Hz
+uint32_t g_tick = 0;
+
+// Ticks since this key was last hit.
+uint8_t g_key_hit[DRIVER_LED_TOTAL];
+
+// Ticks since any key was last hit.
+uint32_t g_any_key_hit = 0;
+
+#ifndef PI
+#define PI 3.14159265
+#endif
+
+uint32_t eeconfig_read_rgb_matrix(void) {
+  return eeprom_read_dword(EECONFIG_RGB_MATRIX);
+}
+void eeconfig_update_rgb_matrix(uint32_t val) {
+  eeprom_update_dword(EECONFIG_RGB_MATRIX, val);
+}
+void eeconfig_update_rgb_matrix_default(void) {
+  dprintf("eeconfig_update_rgb_matrix_default\n");
+  rgb_matrix_config.enable = 1;
+  rgb_matrix_config.mode = RGB_MATRIX_CYCLE_LEFT_RIGHT;
+  rgb_matrix_config.hue = 0;
+  rgb_matrix_config.sat = 255;
+  rgb_matrix_config.val = 255;
+  eeconfig_update_rgb_matrix(rgb_matrix_config.raw);
+}
+void eeconfig_debug_rgb_matrix(void) {
+  dprintf("rgb_matrix_config eprom\n");
+  dprintf("rgb_matrix_config.enable = %d\n", rgb_matrix_config.enable);
+  dprintf("rgb_matrix_config.mode = %d\n", rgb_matrix_config.mode);
+  dprintf("rgb_matrix_config.hue = %d\n", rgb_matrix_config.hue);
+  dprintf("rgb_matrix_config.sat = %d\n", rgb_matrix_config.sat);
+  dprintf("rgb_matrix_config.val = %d\n", rgb_matrix_config.val);
+}
+
+// Last led hit
+#define LED_HITS_TO_REMEMBER 8
+uint8_t g_last_led_hit[LED_HITS_TO_REMEMBER] = {255};
+uint8_t g_last_led_count = 0;
+
+void map_row_column_to_led( uint8_t row, uint8_t column, uint8_t *led_i, uint8_t *led_count) {
+    rgb_led led;
+    *led_count = 0;
+
+    for (uint8_t i = 0; i < DRIVER_LED_TOTAL; i++) {
+        // map_index_to_led(i, &led);
+        led = g_rgb_leds[i];
+        if (row == led.matrix_co.row && column == led.matrix_co.col) {
+            led_i[*led_count] = i;
+            (*led_count)++;
+        }
+    }
+}
+
+
+void rgb_matrix_update_pwm_buffers(void) {
+    IS31FL3731_update_pwm_buffers( DRIVER_ADDR_1, DRIVER_ADDR_2 );
+    IS31FL3731_update_led_control_registers( DRIVER_ADDR_1, DRIVER_ADDR_2 );
+}
+
+void rgb_matrix_set_color( int index, uint8_t red, uint8_t green, uint8_t blue ) {
+    IS31FL3731_set_color( index, red, green, blue );
+}
+
+void rgb_matrix_set_color_all( uint8_t red, uint8_t green, uint8_t blue ) {
+    IS31FL3731_set_color_all( red, green, blue );
+}
+
+
+bool process_rgb_matrix(uint16_t keycode, keyrecord_t *record) {
+    if ( record->event.pressed ) {
+        uint8_t led[8], led_count;
+        map_row_column_to_led(record->event.key.row, record->event.key.col, led, &led_count);
```

Comment:
As far as I see, `process_rgb_matrix` is called every cycle, right? In that case... this part here may become a bit costy, because `map_row_column_to_led` iterates over all `DRIVER_LED_TOTAL` LEDs, for each key pressed. When chording or typing fast, this may have a significant impact on performance.

Would it not be better to have an array that maps rows and columns directly to LED indexes? One could do it with two lookups then: one row,column to LED index, and a LED index to that `led[8]` thing... That'd be a bit more RAM used, but O(1) lookups.

Reply:
My original code had a map of row/col to LED indices... this can be a future optimization.